### PR TITLE
Forbid using derivative functions outside the fragment/light in shader

### LINKED ISSUE #76155: dFdx and dFdy Bricking Spatial Shaders from Vertex Function
### Godot version

4.0 stable

### System information

Windows 10

### Issue description

When the functions `dFdx` and `dFdy` are used in the vertex function of a spatial shader, the shader stops working (as though there were a syntax error), even though no errors are thrown in the shader editor.

The standard editor output does show the following error:
`servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:171 - Condition "!shader_singleton->shader.version_is_valid(version)" is true.`

And many errors are thrown when the project is run.

### Steps to reproduce

From my experimentation, having either of the functions anywhere in the vertex function have the described effect.

### Minimal reproduction project

[dFdx + dFdy Demonstration.zip](https://github.com/godotengine/godot/files/11246120/dFdx.%2B.dFdy.Demonstration.zip)

### PR DESCRIPTION
**Pull request by Chaosus**
Closes https://github.com/godotengine/godot/issues/76155

### CODE REVIEW

File: servers/rendering/shader_language.cpp [hunk 1]
```diff
@@ -10276,11 +10302,33 @@ Error ShaderLanguage::complete(const String &p_code, const ShaderCompileInfo &p_
...
 						idx++;
 						continue;
 					}
+
+					if (is_non_frag_light) {
+						bool found = false;
+						{
+							int idx2 = 0;
+							while (builtin_func_fragment_light_only[idx2].name) {
+								if (builtin_func_fragment_light_only[idx2].name == builtin_func_defs[idx].name) {
```
Comment (AThousandShips) [inline servers/rendering/shader_language.cpp:10318]: This compares `const char *`, are they guaranteed to be the exact same pointer?

Comment (Chaosus) [reply servers/rendering/shader_language.cpp:10318]: Hmm, I didn't find the information about it, but such comparisons seem valid in C++:

изображение

Comment (AThousandShips) [reply servers/rendering/shader_language.cpp:10318]: As long as they're guaranteed to be the same constant, unsure how it works with different translation units and such

**Comment by Chaosus** [comment]
I've noticed it will break compatibility with custom user functions, eg:
```
vec2 test() {
	return fwidth(vec2(0, 1)); // No longer possible.
}

void fragment() {
	ALBEDO.rg = test();
}
```
I think it needs to have a better solution like detection which function declared before vertex function and then forbid using derivative for them (because they may be used inside vertex function which lead to that problem.

**Pull request closed without merging**